File: Dot examples/ASP Book Examples/Book Examples/Chapter18/CS/EntityBaseClasses/EntityBase.cs
using System;
using System.Web;
using System.Web.Caching;
using System.Configuration;
using System.Web.Configuration;
using System.Linq;
using System.Data.Linq;
using System.Data.Linq.Mapping;
using System.Collections.Generic;
using System.Linq.Expressions;
using System.Reflection;

/// <summary>
/// Provides the base class for all entities. Adds Select, Insert, Update, Get and
/// Delete methods to entities including support for data source paging, sorting, and caching
/// </summary>
public abstract class EntityBase
{

    private ValidationErrorCollection _ValidationErrors;


    public ValidationErrorCollection ValidationErrors
    {
        get 
        { 
            if (_ValidationErrors == null)
                _ValidationErrors = new ValidationErrorCollection();
            return _ValidationErrors; 
        }
        set 
        { 
            if (_ValidationErrors == null)
                _ValidationErrors = new ValidationErrorCollection();
            _ValidationErrors = value; 
        }
    }

    public bool IsValid 
    { 
        get
        {
            return (ValidationErrors.Count == 0);
        } 
    }


    protected void PerformValidation()
    {
        Validate();
        if (!IsValid)
            throw new ValidationException(this.ToString() + " failed validation", this);
    }


    protected abstract void Validate();
}


public abstract class EntityBase<T>:EntityBase where T: EntityBase<T>, new()
{


#region Select Methods

    /// <summary>
    /// Returns unsorted list of entity
    /// </summary>
    public  static IQueryable<T> Select()
    {
        return Table;
    }

    /// <summary>
    /// Returns cached unsorted list of entity
    /// </summary>
    public  static IEnumerable<T> SelectCached()
    {
        string key = GetCacheKey("SelectCached");
        List<T> results = (List<T>)HttpContext.Current.Cache[key];
        if (results == null)
        {
            results = Select().ToList();
            AddCache(key, results);
            HttpContext.Current.Trace.Warn("Retrieving " + typeof(T) + " records from database");
        }
        return results;
    }


    /// <summary>
    /// Returns sorted list of entity
    /// </summary>
    /// <param name="orderBy">Name of property to order by</param>
    public  static IEnumerable<T> Select(string orderBy)
    {
        if (String.IsNullOrEmpty(orderBy))
            return Select();

        return SelectDynamicSort(orderBy);
    }


    /// <summary>
    /// Returns unsorted list of paged entities
    /// </summary>
    /// <param name="startRowIndex">The starting database row</param>
    /// <param name="maximumRows">The number of rows to return</param>
    public  static IEnumerable<T> Select(int startRowIndex, int maximumRows)
    {
        if (startRowIndex == -1)
            return new List<T>() { new T() };

        return Select().Skip(startRowIndex).Take(maximumRows);
    }

    /// <summary>
    /// Returns sorted list of paged entities
    /// </summary>
    /// <param name="startRowIndex">The starting database row</param>
    /// <param name="maximumRows">The number of rows to return</param>
    /// <param name="orderBy">The property to order by</param>
    public  static IEnumerable<T> Select(int startRowIndex, int maximumRows, string orderBy)
    {
        if (startRowIndex == -1)
            return new List<T>() { new T() };

        if (String.IsNullOrEmpty(orderBy))
            return Select(startRowIndex, maximumRows);

        return SelectDynamicSort(orderBy).Skip(startRowIndex).Take(maximumRows);
    }



#endregion


#region Sort Methods

    /// <summary>
    /// Takes an expression like AuthorName DESC and breaks it
    /// into AuthorName and DESC
    /// </summary>
    /// <param name="orderBy">The full order by expression</param>
    /// <param name="orderByColumn">The name of the order by column</param>
    /// <param name="orderByDirection">The direction of the sort (asc or desc)</param>
    private static void ParseOrderBy(string orderBy, ref string orderByColumn, ref string orderByDirection)
    {
        string[] orderByParts = orderBy.Split(' ');
        orderByColumn = orderByParts[0];
        if (orderByParts.Length > 1)
            orderByDirection = orderByParts[1].ToLower();
    }


    /// <summary>
    /// Used to perform typed and dynamic order by.
    /// </summary>
    /// <param name="orderBy">The full order by expression, e.g. AuthorName DESC</param>
    private static IEnumerable<T> SelectDynamicSort(string orderBy)
    {
        // Parse order by            
        string orderByColumn = String.Empty;
        string orderByDirection = "asc";
        ParseOrderBy(orderBy, ref orderByColumn, ref orderByDirection);

        // Get sort results
        PropertyInfo prop = typeof(T).GetProperty(orderByColumn);
        switch (Type.GetTypeCode(prop.PropertyType))
        {
            case TypeCode.Boolean:
                return GetDynamicSortResults<bool>(orderByColumn, orderByDirection);
            case TypeCode.Byte:
                return GetDynamicSortResults<byte>(orderByColumn, orderByDirection);
            case TypeCode.Char:
                return GetDynamicSortResults<char>(orderByColumn, orderByDirection);
            case TypeCode.DateTime:
                return GetDynamicSortResults<DateTime>(orderByColumn, orderByDirection);
            case TypeCode.Decimal:
                return GetDynamicSortResults<Decimal>(orderByColumn, orderByDirection);
            case TypeCode.Double:
                return GetDynamicSortResults<double>(orderByColumn, orderByDirection);
            case TypeCode.Int16:
                return GetDynamicSortResults<Int16>(orderByColumn, orderByDirection);
            case TypeCode.Int32:
                return GetDynamicSortResults<Int32>(orderByColumn, orderByDirection);
            case TypeCode.Int64:
                return GetDynamicSortResults<Int64>(orderByColumn, orderByDirection);
            case TypeCode.Single:
                return GetDynamicSortResults<Single>(orderByColumn, orderByDirection);
            case TypeCode.String:
                return GetDynamicSortResults<String>(orderByColumn, orderByDirection);
            default:
                throw new Exception("Cannot sort column " + orderByColumn + " because of its type");
        }
    }


    /// <summary>
    /// Dynamically builds a Linq expression that sorts by a particular column
    /// </summary>
    /// <typeparam name="S">The return type of the Lambda</typeparam>
    /// <param name="orderByColumn">The name of the order by column</param>
    /// <param name="orderByDirection">The direction of the order by sort</param>
    private static IEnumerable<T> GetDynamicSortResults<S>(string orderByColumn, string orderByDirection)
    {
        ParameterExpression sortParameterExpression = Expression.Parameter(typeof(T), "e");
        PropertyInfo propInfo = typeof(T).GetProperty(orderByColumn);
        MemberExpression sortPropertyExpression = Expression.MakeMemberAccess(sortParameterExpression, propInfo);

        Expression<Func<T, S>> sortExpression = Expression.Lambda<Func<T, S>>(sortPropertyExpression, sortParameterExpression);

        if (orderByDirection == "asc")
            return Table.OrderBy(sortExpression);
        else
            return Table.OrderByDescending(sortExpression);
    }


#endregion



#region Select Count Methods


    /// <summary>
    /// Returns count of the records in a database
    /// </summary>
    public static int SelectCount()
    {
        return Table.Count();
    }

    /// <summary>
    /// Returns count of the records in a database 
    /// </summary>
    /// <param name="orderBy">The order by </param>
    public static int SelectCount(string orderBy)
    {
        return SelectCount();
    }



    /// <summary>
    /// Returns count of the records in a database, caching the
    /// result.
    /// </summary>
    public static int SelectCountCached()
    {
        string key = GetCacheKey("Count");
        int? count = (int?)HttpContext.Current.Cache[key];
        if (count == null)
        {
            count = SelectCount();
            AddCache(key, count);
        }
        return count.Value;
    }


    /// <summary>
    /// Returns count of records in database
    /// </summary>
    /// <param name="orderBy">The orderby parameter is ignored, but 
    /// required by ObjectDataSource control</param>
    public  static int SelectCountCached(string orderBy)
    {
        return SelectCountCached();
    }


#endregion

   
#region Get Methods   


    public static T Get(int? id)
    {
        if (id == null)
            return new T();
        
        return Table.Single(GetDynamicGet(id.Value));    
    }


    private static Expression<Func<T, bool>> GetDynamicGet(int id)
    {
        ParameterExpression e = Expression.Parameter(typeof(T), "e");
        string identityColumnName = GetIdentityColumnName();
        PropertyInfo propInfo = typeof(T).GetProperty(identityColumnName);
        MemberExpression m = Expression.MakeMemberAccess(e, propInfo);
        ConstantExpression c = Expression.Constant(id, typeof(int));
        BinaryExpression b = Expression.Equal(m, c);
        Expression<Func<T, bool>> lambda = Expression.Lambda<Func<T, bool>>(b, e);
        return lambda;
    }
   
   
   private static string GetIdentityColumnName()
   {
        return Context.Mapping.GetMetaType(typeof(T)).DBGeneratedIdentityMember.Name;
   }
   
   

#endregion


#region Save Methods

   public virtual void Save(T newEntity)
   {
       T oldEntity = (T)this;
       string identityColumnName = GetIdentityColumnName();
       if ((int)oldEntity.GetType().GetProperty(identityColumnName).GetValue(oldEntity, null) == 0)
           newEntity.Insert();
       else
           oldEntity.Update(newEntity);
   }


    public static T Save(T oldEntity, T newEntity)
    {
        oldEntity.Save(newEntity);
        return newEntity;
    }

#endregion


#region Insert Methods


    public virtual void Insert()
    {
        this.PerformValidation();
        Table.InsertOnSubmit((T)this);
        Context.SubmitChanges();
        ClearCache();

        if (this == null)
            throw new Exception(String.Format("Could not insert {0}", typeof(T)));
    }

    public static T Insert(T entityToInsert)
    {
        entityToInsert.Insert();
        return entityToInsert;
    }



#endregion


#region Update Methods

    public virtual void Update(T newEntity)
    {
        T oldEntity = (T)this;
        newEntity.PerformValidation();
        Table.Attach(newEntity, oldEntity);
        Context.SubmitChanges();
        ClearCache();

        if (oldEntity == null)
            throw new Exception(String.Format("Could not update {0}", typeof(T)));

        oldEntity.ValidationErrors = newEntity.ValidationErrors;
        newEntity = oldEntity;
    }


    public static T Update(T oldEntity, T newEntity)
    {
        oldEntity.Update(newEntity);
        return oldEntity;
    }



    
#endregion



#region Delete Methods


    public virtual void Delete()
    {
        T entityToDelete = (T)this;
        Table.Attach(entityToDelete);
        Table.DeleteOnSubmit(entityToDelete);
        Context.SubmitChanges();
        ClearCache();
    }


    public static T Delete(T entityToDelete)
    {
        entityToDelete.Delete();
        return entityToDelete;
    }

    
#endregion    
    


#region Cache Methods


    protected static void AddCache(string key, object value)
    {
        EnsurePrimaryCacheKey();
        string primaryKey = GetPrimaryCacheKey();
        HttpContext context = HttpContext.Current;
        CacheDependency keyDepend = new CacheDependency(null, new string[] { primaryKey });
        context.Cache.Insert(key, value, keyDepend);

    }


    private static void EnsurePrimaryCacheKey()
    {
        string primaryKey = GetPrimaryCacheKey();
        HttpContext context = HttpContext.Current;
        if (context.Cache[primaryKey] == null)
            context.Cache.Insert(primaryKey, DateTime.Now.ToString());
    }



    private static string GetPrimaryCacheKey()
    {
        return String.Format("{0}_Cache", typeof(T));
    }


    public  static string GetCacheKey(string key)
    {
        return String.Format("{0}_{1}", GetPrimaryCacheKey(), key);
    }


    public static void ClearCache()
    {
        HttpContext.Current.Cache.Remove(GetPrimaryCacheKey());
    }



#endregion




    public static Table<T> Table
    {
        get
        {
            return Context.GetTable<T>();
        }
    }



    protected static DataContext Context
    {
        get
        {
            DataContext context = (DataContext)HttpContext.Current.Items["DataContext"];
            if (context == null)
            {
                ConnectionStringSettings conSettings = WebConfigurationManager.ConnectionStrings["con"];
                if (conSettings == null)
                    throw new Exception("Missing connection string named 'con' in web configuration.");
                context = new DataContext(conSettings.ConnectionString);
                HttpContext.Current.Items["DataContext"] = context;
                context.Log = new TraceWriter();
            }   
            return context; 
        }
    }


    


}

File: Dot examples/ASP Book Examples/Book Examples/Chapter34/vb/App_Code/EntityBaseClasses/EntityBase.vb
Imports System
Imports System.Web
Imports System.Web.Caching
Imports System.Configuration
Imports System.Web.Configuration
Imports System.Linq
Imports System.Data.Linq
Imports System.Data.Linq.Mapping
Imports System.Collections.Generic
Imports System.Linq.Expressions
Imports System.Reflection

''' <summary>
''' Provides the base class for all entities. Adds Select, Insert, Update, Get and
''' Delete methods to entities including support for data source paging, sorting, and caching
''' </summary>
Public MustInherit Class EntityBase


    Private _ValidationErrors As ValidationErrorCollection


    Public Property ValidationErrors() As ValidationErrorCollection
        Get
            If _ValidationErrors Is Nothing Then
                _ValidationErrors = New ValidationErrorCollection()
            End If
            Return _ValidationErrors
        End Get
        set 
            If _ValidationErrors Is Nothing Then
                _ValidationErrors = New ValidationErrorCollection()
            End If
            _ValidationErrors = Value
        End Set
    End Property

    Public ReadOnly Property IsValid() As Boolean
        Get
            Return ValidationErrors.Count = 0
        End Get
    End Property


    Protected Sub PerformValidation()
        Validate()
        If Not IsValid Then
            Throw New ValidationException(Me.ToString() + " failed validation", Me)
        End If
    End Sub

    Protected MustOverride Sub Validate()

end class



Public MustInherit Class EntityBase(Of T As {EntityBase(Of T), New})
    Inherits EntityBase



#Region "Select Methods"

    ''' <summary>
    ''' Returns unsorted list of entity
    ''' </summary>
    Public Shared Function [Select]() As IQueryable(Of T)
        Return Table
    End Function

    ''' <summary>
    ''' Returns cached unsorted list of entity
    ''' </summary>
    Public Shared Function SelectCached() As IEnumerable(Of T)
        Dim key As String = GetCacheKey("SelectCached")
        Dim results As List(Of T) = CType(HttpContext.Current.Cache(key), List(Of T))
        If results Is Nothing Then
            results = [Select]().ToList()
            AddCache(key, results)
            HttpContext.Current.Trace.Warn("Retrieving " & GetType(T).ToString() & " records from database")
        End If
        Return results
    End Function


    ''' <summary>
    ''' Returns sorted list of entity
    ''' </summary>
    ''' <param name="orderBy">Name of property to order by</param>
    Public Shared Function [Select](ByVal orderBy As String) As IEnumerable(Of T)
        If String.IsNullOrEmpty(orderBy) Then
            Return [Select]()
        End If
        Return SelectDynamicSort(orderBy)
    End Function


    ''' <summary>
    ''' Returns unsorted list of paged entities
    ''' </summary>
    ''' <param name="startRowIndex">The starting database row</param>
    ''' <param name="maximumRows">The number of rows to return</param>
    Public Shared Function [Select](ByVal startRowIndex As Integer, ByVal maximumRows As Integer) As IEnumerable(Of T)
        If startRowIndex = -1 Then
            Dim result = New List(Of T)
            result.Add(New T())
            Return result
        End If
        Return [Select]().Skip(startRowIndex).Take(maximumRows)
    End Function

    ''' <summary>
    ''' Returns sorted list of paged entities
    ''' </summary>
    ''' <param name="startRowIndex">The starting database row</param>
    ''' <param name="maximumRows">The number of rows to return</param>
    ''' <param name="orderBy">The property to order by</param>
    Public Shared Function [Select](ByVal startRowIndex As Integer, ByVal maximumRows As Integer, ByVal orderBy As String) As IEnumerable(Of T)
        If startRowIndex = -1 Then
            Dim result = New List(Of T)
            result.Add(New T())
            Return result
        End If

        If String.IsNullOrEmpty(orderBy) Then
            Return [Select](startRowIndex, maximumRows)
        End If

        Return SelectDynamicSort(orderBy).Skip(startRowIndex).Take(maximumRows)
    End Function



#End Region



#Region "Sort Methods"

    ''' <summary>
    ''' Takes an expression like AuthorName DESC and breaks it
    ''' into AuthorName and DESC
    ''' </summary>
    ''' <param name="orderBy">The full order by expression</param>
    ''' <param name="orderByColumn">The name of the order by column</param>
    ''' <param name="orderByDirection">The direction of the sort (asc or desc)</param>
    Private Shared Sub ParseOrderBy(ByVal orderBy As String, ByRef orderByColumn As String, ByRef orderByDirection As String)
        Dim orderByParts As String() = orderBy.Split(" ")
        orderByColumn = orderByParts(0)
        If orderByParts.Length > 1 Then
            orderByDirection = orderByParts(1).ToLower()
        End If
    End Sub


    ''' <summary>
    ''' Used to perform typed and dynamic order by.
    ''' </summary>
    ''' <param name="orderBy">The full order by expression, e.g. AuthorName DESC</param>
    Private Shared Function SelectDynamicSort(ByVal orderBy As String) As IEnumerable(Of T)
        ' Parse order by            
        Dim orderByColumn As String = String.Empty
        Dim orderByDirection As String = "asc"
        ParseOrderBy(orderBy, orderByColumn, orderByDirection)

        ' Get sort results
        Dim prop As PropertyInfo = GetType(T).GetProperty(orderByColumn)
        Select Type.GetTypeCode(prop.PropertyType)
            Case TypeCode.Boolean
                Return GetDynamicSortResults(Of Boolean)(orderByColumn, orderByDirection)
            Case TypeCode.Byte
                Return GetDynamicSortResults(Of Byte)(orderByColumn, orderByDirection)
            Case TypeCode.Char
                Return GetDynamicSortResults(Of Char)(orderByColumn, orderByDirection)
            Case TypeCode.DateTime
                Return GetDynamicSortResults(Of DateTime)(orderByColumn, orderByDirection)
            Case TypeCode.Decimal
                Return GetDynamicSortResults(Of Decimal)(orderByColumn, orderByDirection)
            Case TypeCode.Double
                Return GetDynamicSortResults(Of Double)(orderByColumn, orderByDirection)
            Case TypeCode.Int16
                Return GetDynamicSortResults(Of Int16)(orderByColumn, orderByDirection)
            Case TypeCode.Int32
                Return GetDynamicSortResults(Of Int32)(orderByColumn, orderByDirection)
            Case TypeCode.Int64
                Return GetDynamicSortResults(Of Int64)(orderByColumn, orderByDirection)
            Case TypeCode.Single
                Return GetDynamicSortResults(Of Single)(orderByColumn, orderByDirection)
            Case TypeCode.String
                Return GetDynamicSortResults(Of String)(orderByColumn, orderByDirection)
            Case Else
                Throw New Exception("Cannot sort column " & orderByColumn & " because of its type")
        End Select
    End Function


    ''' <summary>
    ''' Dynamically builds a Linq expression that sorts by a particular column
    ''' </summary>
    ''' <typeparam name="S">The return type of the Lambda</typeparam>
    ''' <param name="orderByColumn">The name of the order by column</param>
    ''' <param name="orderByDirection">The direction of the order by sort</param>
    Private Shared Function GetDynamicSortResults(Of S)(ByVal orderByColumn As String, ByVal orderByDirection As String) As IEnumerable(Of T)
        Dim sortParameterExpression As ParameterExpression = Expression.Parameter(GetType(T), "e")
        Dim propInfo As PropertyInfo = GetType(T).GetProperty(orderByColumn)
        Dim sortPropertyExpression As MemberExpression = Expression.MakeMemberAccess(sortParameterExpression, propInfo)

        Dim sortExpression = Expression.Lambda(Of Func(Of T, S))(sortPropertyExpression, sortParameterExpression)

        If orderByDirection = "asc" Then
            Return Table.OrderBy(sortExpression)
        Else
            Return Table.OrderByDescending(sortExpression)
        End If
    End Function


#End Region



#Region "Select Count Methods"


    ''' <summary>
    ''' Returns count of the records in a database
    ''' </summary>
    Public Shared Function SelectCount() As Integer
        Return Table.Count()
    End Function

    ''' <summary>
    ''' Returns count of the records in a database 
    ''' </summary>
    ''' <param name="orderBy">The order by </param>
    Public Shared Function SelectCount(ByVal orderBy As String) As Integer
        Return SelectCount()
    End Function



    ''' <summary>
    ''' Returns count of the records in a database, caching the
    ''' result.
    ''' </summary>
    Public Shared Function SelectCountCached() As Integer
        Dim key As String = GetCacheKey("Count")
        Dim count As Nullable(Of Integer) = CType(HttpContext.Current.Cache(key), Nullable(Of Integer))
        If count Is Nothing Then
            count = SelectCount()
            AddCache(key, count)
        End If
        Return count.Value
    End Function


    ''' <summary>
    ''' Returns count of records in database
    ''' </summary>
    ''' <param name="orderBy">The orderby parameter is ignored, but 
    ''' required by ObjectDataSource control</param>
    Public Shared Function SelectCountCached(ByVal orderBy As String) As Integer
        Return SelectCountCached()
    End Function


#End Region


#Region "Get Methods"


    Public Shared Function [Get](ByVal id As Nullable(Of Integer)) As T
        If IsNothing(Id) Then
            Return New T()
        End If
        Return Table.Single(GetDynamicGet(id.Value))
    End Function


    Private Shared Function GetDynamicGet(ByVal id As Integer) As Expression(Of Func(Of T, Boolean))
        Dim e As ParameterExpression = Expression.Parameter(GetType(T), "e")
        Dim identityColumnName As String = GetIdentityColumnName()
        Dim propInfo As PropertyInfo = GetType(T).GetProperty(identityColumnName)
        Dim m As MemberExpression = Expression.MakeMemberAccess(e, propInfo)
        Dim c As ConstantExpression = Expression.Constant(id, GetType(Integer))
        Dim b As BinaryExpression = Expression.Equal(m, c)
        Dim lambda = Expression.Lambda(Of Func(Of T, Boolean))(b, e)
        Return lambda
    End Function


   Private Shared Function GetIdentityColumnName() As String
        Return Context.Mapping.GetMetaType(GetType(T)).DBGeneratedIdentityMember.Name
   End Function



#End Region


#Region "Save Methods"

   Public Overridable Sub Save(ByVal newEntity As T)
       Dim oldEntity As T = DirectCast(Me, T)
       Dim identityColumnName As String = GetIdentityColumnName()
       If oldEntity.GetType().GetProperty(identityColumnName).GetValue(oldEntity, Nothing) = 0 Then
           newEntity.Insert()
       Else
           oldEntity.Update(newEntity)
        End If
   End Sub


    Public Shared Function Save(ByVal oldEntity As T, ByVal newEntity As T) As T
        oldEntity.Save(newEntity)
        Return newEntity
    End Function

#End Region


#Region "Insert Methods"


    Public Overridable Sub Insert()
        Me.PerformValidation()
        Table.InsertOnSubmit(DirectCast(Me, T))
        Context.SubmitChanges()
        ClearCache()
        If (IsNothing(Me)) Then
            Throw New Exception(String.Format("Could not insert {0}", GetType(T)))
        End If
    End Sub

    Public Shared Function Insert(ByVal entityToInsert As T) As T
        entityToInsert.Insert()
        Return entityToInsert
    End Function



#End Region


#Region "Update Methods"

    Public Overridable Sub Update(ByVal newEntity As T)
        Dim oldEntity As T = DirectCast(Me, T)
        newEntity.PerformValidation()
        Table.Attach(newEntity, oldEntity)
        Context.SubmitChanges()
        ClearCache()

        If IsNothing(oldEntity) Then
            Throw New Exception(String.Format("Could not update {0}", GetType(T)))
        End If

        oldEntity.ValidationErrors = newEntity.ValidationErrors
        newEntity = oldEntity
    End Sub


    Public Shared Function Update(ByVal oldEntity As T, ByVal newEntity As T) As T
        oldEntity.Update(newEntity)
        Return oldEntity
    End Function



#End Region



#Region "Delete Methods"


    Public Overridable Sub Delete()
        Dim entityToDelete As T = DirectCast(Me, T)
        Table.Attach(entityToDelete)
        Table.DeleteOnSubmit(entityToDelete)
        Context.SubmitChanges()
        ClearCache()
    End Sub


    Public Shared Function Delete(ByVal entityToDelete As T) As T
        entityToDelete.Delete()
        Return entityToDelete
    End Function


#End Region



#Region "Cache Methods"


    Protected Shared Sub AddCache(ByVal key As String, ByVal value As Object)
        EnsurePrimaryCacheKey()
        Dim primaryKey As String = GetPrimaryCacheKey()
        Dim context As HttpContext = HttpContext.Current
        Dim keyDepend As New CacheDependency(Nothing, New String() {primaryKey})
        context.Cache.Insert(key, value, keyDepend)
    End Sub


    Private Shared Sub EnsurePrimaryCacheKey()
        Dim primaryKey As String = GetPrimaryCacheKey()
        Dim context As HttpContext = HttpContext.Current
        If IsNothing(context.Cache(primaryKey)) Then
            context.Cache.Insert(primaryKey, DateTime.Now.ToString())
        End If
    End Sub



    Private Shared Function GetPrimaryCacheKey() As String
        Return String.Format("{0}_Cache", GetType(T))
    End Function


    Public Shared Function GetCacheKey(ByVal key As String) As String
        Return String.Format("{0}_{1}", GetPrimaryCacheKey(), key)
    End Function


    Public Shared Sub ClearCache()
        HttpContext.Current.Cache.Remove(GetPrimaryCacheKey())
    End Sub



#End Region




    Public Shared ReadOnly Property Table() As Table(Of T)
        Get
            Return Context.GetTable(Of T)()
        End Get
    End Property



    Protected Shared ReadOnly Property Context() As DataContext
        Get
            Dim myContext As DataContext = DirectCast(HttpContext.Current.Items("DataContext"), DataContext)
            If IsNothing(myContext) Then
                Dim conSettings As ConnectionStringSettings = WebConfigurationManager.ConnectionStrings("con")
                If IsNothing(conSettings) Then
                    Throw New Exception("Missing connection string named 'con' in web configuration.")
                End If
                myContext = New DataContext(conSettings.ConnectionString)
                HttpContext.Current.Items("DataContext") = myContext
                myContext.Log = New TraceWriter()
            End If
            Return myContext
        End Get
    End Property





End Class

